gateway/middleware.go

```diff
@@ -431,6 +435,55 @@ func (t *BaseMiddleware) ApplyPolicies(session *user.SessionState) error {
 	return store.Apply(session)
 }
 
+// recordAccessLog is only used for Success/Error handler
+func (t *BaseMiddleware) recordAccessLog(req *http.Request, resp *http.Response, latency *analytics.Latency) {
+	hashKeys := t.Gw.GetConfig().HashKeys
+	accessLog := accesslog.NewRecord()
+	allowedFields := t.Gw.GetConfig().AccessLogs.Template
+	fields := accessLog.Fields()
+
+	// Set the access log fields
+	accessLog.WithApiKey(req, hashKeys, t.Gw.obfuscateKey)
+	accessLog.WithClientIP(req)
+	accessLog.WithRemoteAddr(req)
+	accessLog.WithHost(req)
+	accessLog.WithLatencyTotal(latency)
+	accessLog.WithMethod(req)
+	accessLog.WithPath(req)
+	accessLog.WithProtocol(req)
+	accessLog.WithStatus(resp)
+	accessLog.WithUpstreamAddress(req)
+	accessLog.WithUpstreamLatency(latency)
+	accessLog.WithUserAgent(req)
```

Comment:
You don't need all these functions, you only need:

- WithKey (api in name is redundant)
- WithRequest(req)
- WithResponse(resp)
- WithLatency(latency)

Everything else should be done inside, no need for the large API footprint. Makes the set fields keys easier to review.